tests/append_test.go

```diff
@@ -261,3 +261,42 @@ func TestAppendConnectorReturnsSameRemoteIDDifferentMBox(t *testing.T) {
 		}
 	})
 }
+
+func TestAppendCanHandleOutOfOrderUIDUpdates(t *testing.T) {
+	// Make sure we are correctly handling the case where we have to clients doing append at the same time.
+	// Both clients append a message and get assigned UID according to whichever got there first:
+	//
+	// * Client A -> Append -> UID 1
+	// * Client B -> Append -> UID 2
+	//
+	// All Clients apply their changes to their local state immediately and will receive a deferred updates for the
+	// same mailbox if other clients make updates.
+	// In the case of client B, it appends UID2 as the first message and then later receives an update from A with
+	// an UID lower than the last UID which causes a panic.
+	//
+	runManyToOneTestWithAuth(t, defaultServerOptions(t), []int{1, 2}, func(c map[int]*testConnection, session *testSession) {
+		const MessageCount = 20
+
+		// Select mailbox so that both clients get updates.
+		c[1].C("A001 SELECT INBOX").OK("A001")
+		c[2].C("A002 SELECT INBOX").OK("A002")
+
+		appendFN := func(clientIndex int) {
+			for i := 0; i < MessageCount; i++ {
+				c[clientIndex+1].doAppend("INBOX", "To: f3@pm.me\r\n", "\\Seen").expect("OK")
+			}
+		}
+
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+
+		for i := 0; i < 2; i++ {
+			go func(index int) {
+				defer wg.Done()
+				appendFN(index)
+			}(i)
+		}
+
+		wg.Wait()
```

Comment:
at this point, after clients have received all updates from each other, could we perform a list with each client and assert that they all see the same messages in the same order?

Reply:
Good point. Fixed. I also had to add option to disable parallelism to make sure the fetch results arrive in order.